primitives/state-machine/src/trie_backend.rs

```diff
@@ -184,15 +187,41 @@ where
 	}
 }
 
+/// A cached iterator.
+struct CachedIter<S, H, C>
+where
+	H: Hasher,
+{
+	is_cached: bool,
+	last_key: sp_std::vec::Vec<u8>,
+	iter: RawIter<S, H, C>,
+}
+
+impl<S, H, C> Default for CachedIter<S, H, C>
+where
+	H: Hasher,
+{
+	fn default() -> Self {
+		Self { is_cached: false, last_key: Default::default(), iter: Default::default() }
+	}
+}
+
 /// Patricia trie-based backend. Transaction type is an overlay of changes to commit.
 pub struct TrieBackend<S: TrieBackendStorage<H>, H: Hasher, C = LocalTrieCache<H>> {
 	pub(crate) essence: TrieBackendEssence<S, H, C>,
+	#[cfg(feature = "std")]
+	next_storage_key_cache: Mutex<CachedIter<S, H, C>>,
```

Comment:
would make sense to have it for no_std (behind refcell).
Why not `Mutex<Option<CachedIter<S, H, C>>`? (and rem `is_cached`).

Reply:
Done.